File: example/python/xpt/tcp_client.py
"""Portable XPT TCP client: connect, send, receive, peer check, disconnect.

XPT provides no listener API, so the peer comes from the command line; point it at any TCP echo
service. Without a host argument, or without a server listening there, the sample reports that in
one sentence and exits cleanly. Build with WSE_BUILD_XPT=ON and WSE_BUILD_PYTHON_BINDING=ON.

XPT deliberately has no default timeout: every operation takes an explicit OperationContext so a
caller can never wait forever by accident.

Both arguments are required, and the module path is the first of them, so this sample always loads
the built extension rather than the installed `wse` package.

    python example/python/xpt/tcp_client.py <path to the built _wse module> <host> <port>
"""

import importlib.util
import sys
from pathlib import Path

# One second is generous for a connect on a local network and short enough that pointing the sample
# at an unreachable host still returns promptly. A production client would size this from the route
# it expects rather than reuse one number for connect and receive alike.
TIMEOUT_MS = 1000
# The sample prints what came back, so the limit only has to hold a short echo. It is a ceiling on
# one receive, not a buffer that is always filled: a smaller reply returns as soon as it arrives.
RECEIVE_LIMIT = 256


def load_wse():
    """Loads the binding, either as the installed package or from a built module path.

    With no argument the installed `wse` package is imported. With one, the built `_wse` module at
    that path is loaded directly, which is how the sample runs against a build tree. main() below
    requires the path, so only the second mode is reachable here.
    """
    if len(sys.argv) == 1:
        import wse

        return wse
    module_path = Path(sys.argv[1]).resolve()
    spec = importlib.util.spec_from_file_location("_wse", module_path)
    if spec is None or spec.loader is None:
        raise RuntimeError(f"Unable to load WSE Python module: {module_path}")
    module = importlib.util.module_from_spec(spec)
    sys.modules[spec.name] = module
    spec.loader.exec_module(module)
    return module


def report_transport_error(operation: str, failure) -> None:
    """Prints one structured line for a WseError: the operation, then the whole error identity.

    Every field is printed because they answer different questions. `category` says which layer
    refused, `code` is the portable TransportErrorCode a caller may branch on, `native_code` is the
    untranslated errno or WSAGetLastError value that a platform-specific diagnosis needs, and the
    message is prose for a human. A caller branches on category and code, never on the message.
    """
    print(
        f"ERROR: {operation} failed. category={failure.category} code={failure.code}"
        f" native={failure.native_code} message={failure}"
    )


def main() -> int:
    # Argument handling first, before the binding is loaded: a usage mistake should not depend on
    # whether the native module is present. Every bad-argument path prints and returns 0, because
    # running the sample wrongly is not a build failure.
    if len(sys.argv) < 4:
        print(
            "No host was given; pass the built _wse module path, then a host and a port:"
            " python example/python/xpt/tcp_client.py <module path> <host> <port>"
        )
        return 0
    host = sys.argv[2]
    try:
        port = int(sys.argv[3])
    except ValueError:
        print(f"The port argument is not a number: {sys.argv[3]}")
        return 0

    wse = load_wse()
    # A build without WSE_BUILD_XPT registers no TcpClient at all, so the name has to be looked for
    # rather than caught; there is no stub that raises a friendlier error later.
    if not hasattr(wse, "TcpClient"):
        raise SystemExit("This WSE build does not include the XPT component.")

    # One explicit deadline is reused for every operation in this sample.
    # A context is a plain value and carries no socket state, so sharing one across connect, send,
    # and receive is safe; each call starts its own countdown from that same duration rather than
    # spending one shared budget.
    context = wse.OperationContext(TIMEOUT_MS)

    # The context manager is the single owner; leaving it disconnects the socket.
    with wse.TcpClient() as client:
        try:
            client.connect(host, port, context)
        except wse.WseError as failure:
            # No server on the far end is a reportable state, not a sample defect.
            # CONNECTION_REFUSED, HOST_NOT_FOUND, NETWORK_UNREACHABLE, and TIMED_OUT all land here
            # and all mean the same thing to this sample: nothing is listening where the caller
            # pointed it. Returning 0 is deliberate so the sample runs on a machine with no network
            # at all; a real client would branch on the code and decide whether a retry helps.
            report_transport_error(f"connect {host}:{port}", failure)
            print("point the host and port arguments at a reachable TCP service")
            return 0

        # Both endpoints are printed because only the remote one was supplied by the caller. The
        # local pair is what the operating system chose, and it is the only way to identify this
        # connection in a packet capture or in the peer's own logs.
        local_host, local_port = client.local_endpoint()
        remote_host, remote_port = client.remote_endpoint()
        print(f"connected: {local_host}:{local_port} -> {remote_host}:{remote_port}")

        # send() loops until the complete buffer is written or the deadline expires.
        # That is the asymmetry worth remembering: send is all-or-deadline, while the receive below
        # returns whatever one read produced. A caller that needs a whole message back has to loop.
        print("sent bytes:", client.send(b"WSE-XPT-TCP", context))

        # One receive returns a single chunk; an orderly peer close is REMOTE_CLOSED.
        try:
            payload = client.receive(RECEIVE_LIMIT, context)
            print(f"received {len(payload)} bytes:", payload.decode("utf-8", "replace"))
        except wse.WseError as failure:
            # A silent peer is the normal outcome when the argument points at something that is not
            # an echo service, so a timeout is reported and the sample carries on. Anything else is
            # a genuine transport failure and is printed in full, but still without failing the run.
            if failure.code == wse.TransportErrorCode.TIMED_OUT.value:
                print("no reply within the deadline (the peer is not an echo service)")
            else:
                report_transport_error("receive", failure)

        # A non-destructive peek for an OS-observable peer close or reset.
        # It consumes nothing, so calling it does not disturb a reply still in flight. It can only
        # see what the operating system already knows: a peer that vanished without sending a FIN
        # or a RST is indistinguishable from a healthy idle one until a send fails.
        try:
            client.check_peer_connection()
            print("peer check: no close or reset observed")
        except wse.WseError as failure:
            print("peer check: peer reported code =", failure.code)

        # disconnect() is idempotent; leaving the context manager would do the same.
        # It is written out because the sample wants to print after the socket is down, which is
        # the one thing the context manager cannot express.
        client.disconnect()
        print("disconnected")

    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: example/js/xpt/tcp_client.js
'use strict';

// Portable XPT TCP client: connect, report both endpoints, send, receive, peer check, disconnect.
//
// XPT provides no listener API, so the peer comes from the command line; point it at any TCP echo
// service. Without a reachable server the sample reports the structured connection error and
// returns without failing. Build with WSE_BUILD_XPT=ON and WSE_BUILD_NODE_BINDING=ON.
//
// XPT deliberately has no default timeout: every operation takes an explicit context so a caller
// can never wait forever by accident. These calls block the calling thread until they finish or
// the deadline elapses, so keep the deadline short on the main thread.
//
// Synchronous means synchronous: each call returns a value, not a promise, and nothing else on
// the loop runs while it is in flight. A context may also carry a CancellationSource, but on the
// main thread there is nobody left to call cancel() during the call, so cancellation is only
// useful when the operation runs in a Worker.
//
// The first argument is either the installed `lang/js` package directory or a built `wse.node`
// addon. Given the addon the sample still loads the package, so what is exercised is the public
// JavaScript surface an application uses rather than the raw native entry points.
//
//   node example/js/xpt/tcp_client.js <installed lang/js directory or wse.node> <host> <port>
//
// Argument order: [2] package or addon, [3] host, [4] TCP port.

const path = require('node:path');

const target = process.argv[2] || process.env.WSE_JAVASCRIPT_PACKAGE;
if ( !target ) {
  throw new Error('Pass the installed lang/js directory, or the built wse.node addon.');
}
// A `.node` argument is the build-tree addon. Point the package loader at it and still require the
// package, so the sample always demonstrates the same public surface an application uses.
if ( target.endsWith('.node') ) process.env.WSE_NODE_ADDON = path.resolve(target);
const wse = require(target.endsWith('.node')
  ? path.resolve(__dirname, '..', '..', '..', 'lang', 'js')
  : path.resolve(target));

if ( wse.TcpClient === undefined ) {
  throw new Error('This WSE build does not include the XPT component.');
}

// The ceiling for one receive, not a length to wait for. A stream has no message boundaries, so a
// larger ceiling only means more bytes may arrive at once; whatever is left stays buffered for the
// next receive. Framing a message is the caller's job, not the transport's.
const RECEIVE_LIMIT_BYTES = 256;

async function main() {
  const host = process.argv[3];
  const port = Number(process.argv[4]);
  // The port is range-checked here rather than left to the binding, because a caller who typed a
  // wrong number deserves this message and not an InvalidArgument out of the transport.
  if ( !host || !Number.isInteger(port) || port <= 0 || port > 65535 ) {
    // TCP port 7 is the conventional echo service port.
    // Deliberate: with no peer named there is nothing to connect to, and the sample exits with
    // success so it runs on a machine with no service listening.
    console.log('Pass a host and a TCP port to connect to, for example 127.0.0.1 7.');
    return;
  }

  // One explicit deadline is reused for every operation in this sample.
  // Reused, not shared: each call gets a fresh 1000 ms, and the object carries no state between
  // them. A second is ample on a loopback or LAN service and short enough that an unreachable
  // host does not leave the sample sitting on the main thread; a distant host may need more.
  const context = { timeoutMs: 1000 };

  // The object owns the socket from connect() until disconnect(). JavaScript has no deterministic
  // destructor, so nothing closes that socket on its own: skip disconnect() and it stays open
  // until the garbage collector finalizes this object.
  const client = new wse.TcpClient();
  try {
    client.connect(host, port, context);
  } catch (error) {
    // Returning here rather than entering the try/finally below is safe precisely because a
    // failed connect leaves no socket to disconnect.
    // No server at that address is a reportable state, not a sample defect.
    console.log(`connect ${host}:${port} failed:`,
      error.category, error.code, error.nativeCode, error.message);
    console.log('Point the sample at a reachable TCP service.');
    return;
  }

  try {
    // The local endpoint is only known after a successful connect: the port in it is the ephemeral
    // one the operating system assigned. Both read from the live socket, so both are meaningless
    // once disconnect() has run.
    const local = client.localEndpoint();
    const remote = client.remoteEndpoint();
    console.log(`connected: ${local.host}:${local.port} -> ${remote.host}:${remote.port}`);

    // send() loops until the complete buffer is written or the deadline expires.
    // The returned count is therefore the whole payload on success; a short write is reported as
    // a failure rather than as a smaller number.
    const payload = Buffer.from('wse-xpt-tcp', 'utf8');
    console.log('sent bytes:', client.send(payload, context));

    // One receive returns a single chunk; an orderly peer close is RemoteClosed.
    // Against a peer that is not an echo service, TimedOut is the expected structured result
    // rather than a failure, so it is checked by code and the sample carries on to disconnect
    // cleanly and exit with success.
    try {
      const reply = client.receive(RECEIVE_LIMIT_BYTES, context);
      console.log(`received ${reply.length} bytes:`, reply.toString('utf8'));
    } catch (error) {
      if ( error.code === wse.TransportErrorCode.TimedOut ) {
        console.log('no reply within the deadline (the peer is not an echo service)');
      } else {
        console.log('receive reported:', error.code, error.message);
      }
    }

    // A non-destructive peek for an OS-observable peer close or reset.
    // It consumes nothing and answers immediately, so it takes no context of its own. It reports
    // only what the operating system already knows; a peer that has gone away silently still
    // looks connected until a send or receive proves otherwise.
    try {
      client.checkPeerConnection();
      console.log('peer check: no close or reset observed');
    } catch (error) {
      console.log('peer check: peer reported code', error.code);
    }
  } finally {
    // disconnect() is idempotent.
    // It is also not terminal: the same object may connect() again afterwards. Running it in a
    // finally is what closes the socket at a known moment even when a call above threw.
    client.disconnect();
  }
  console.log('disconnected');
}

main().catch((error) => {
  console.error(error.category, error.code, error.nativeCode, error.message);
  process.exitCode = 1;
});
